### PR TITLE
[14.0][ADD] website_sale_filter_product_brand

### PR DESCRIPTION
**Pull request by dessanhemrayev**
This is module add brands as a filter on shop sidebar

### CODE REVIEW

**Review by francesco-ooops — APPROVED** [review]
Functional review ok!

File: website_sale_product_brand_filter/__manifest__.py [hunk 1]
```diff
@@ -0,0 +1,22 @@
+# Copyright 2020 Advitus MB
+# License LGPL-3.0 or later (https://www.gnu.org/licenses/lgpl-3.0).
```
Comment (GabbasovDinar) [inline website_sale_product_brand_filter/__manifest__.py]: please, check license: there LGPL-3.0 but in the manifest: "license": "AGPL-3",

File: website_sale_product_brand_filter/controllers/website_sale.py [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+        search=None,
+        products=None,
+        search_product=None,
+        category=None,
+    ):
+        domain = []
+        if not products:
+            domain = [("id", "in", selected_brand_ids)]
+        else:
+            if search or category:
```
Comment (GabbasovDinar) [inline website_sale_product_brand_filter/controllers/website_sale.py]:
```suggestion
        elif search or category:
```

File: website_sale_product_brand_filter/controllers/website_sale.py [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+            attrib_values=attrib_values,
+            search_in_description=search_in_description,
+        )
+
+        return domain
+
+    def _remove_extra_brands(self, brands, search_product, attrib_values):
+        search_product_brands = search_product.mapped("product_brand_id")
+        if attrib_values:
+            brands = brands.filtered(lambda b: b.id in search_product_brands.ids)
```
Comment (GabbasovDinar) [inline website_sale_product_brand_filter/controllers/website_sale.py]: You can move mapping inside the condition
```python
        if attrib_values:
            search_product_brands = search_product.mapped("product_brand_id")
            brands = brands.filtered(lambda b: b.id in search_product_brands.ids)
```

File: website_sale_product_brand_filter/controllers/website_sale.py [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+        # parse selected attributes
+        attrib_list = request.httprequest.args.getlist("attrib")
+        attrib_values = res.qcontext["attrib_values"]
+
+        # get filtered products
+        products = res.qcontext["products"]
+        domain = self._get_search_domain_no_brands(
+            search, category, attrib_values, search_in_description=False
+        )
+        search_product = request.env["product.template"].search(domain)
```
Comment (GabbasovDinar) [inline website_sale_product_brand_filter/controllers/website_sale.py]: rename to search_products - it's recordset

File: website_sale_product_brand_filter/controllers/website_sale.py [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+    def _update_domain(self, brands_list, domain):
+        selected_brand_ids = [int(brand) for brand in brands_list]
+        if brands_list:
+            for leaf in domain:
+                if leaf[0] == "product_brand_id":
+                    domain.remove(leaf)
+            domain += [("product_brand_id", "in", selected_brand_ids)]
+        return domain
+
+    def _get_list_brands(self, req):
```
Comment (GabbasovDinar) [inline website_sale_product_brand_filter/controllers/website_sale.py]: I think the name _get_brand_ids is more appropriate

File: website_sale_product_brand_filter/controllers/website_sale.py [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+        for item in ["brand", "brand_ids"]:
+            if req.getlist(item):
+                return req.getlist(item)
+        return []
```
Comment (GabbasovDinar) [inline website_sale_product_brand_filter/controllers/website_sale.py]: return req.getlist("brand") or req.getlist("brand_ids") or []

File: website_sale_product_brand_filter/controllers/website_sale.py [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+
+    def _get_brands(self, domain):
+        return (
+            request.env["product.brand"]
+            .search(domain)
+            .filtered(lambda x: x.products_count > 0)
+        )
```
Comment (GabbasovDinar) [inline website_sale_product_brand_filter/controllers/website_sale.py]: why do you need a separate search method?

File: website_sale_product_brand_filter/controllers/website_sale.py [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+            selected_brand_ids, search, products, search_product, category
+        )
+
+        brands = self._remove_extra_brands(brands, search_product, attrib_values)
+        # assign values for usage in qweb
+        res.qcontext["brands"] = brands
+        res.qcontext["selected_brand_ids"] = selected_brand_ids
+        res.qcontext["attr_valid"] = (
+            search_product.mapped("attribute_line_ids").mapped("value_ids").ids
+        )
```
Comment (GabbasovDinar) [inline website_sale_product_brand_filter/controllers/website_sale.py]:
```suggestion
        res.qcontext.update({
            "brands": brands,
            "selected_brand_ids": selected_brand_ids,
            "attr_valid": search_product.mapped("attribute_line_ids").mapped("value_ids").ids,
        })
```

File: website_sale_product_brand_filter/static/src/js/website_sale_product_brand_filter.js [hunk 9]
```diff
@@ -0,0 +1,29 @@
+odoo.define("website_sale_product_brand_filter.website", function () {
+    "use strict";
+
+    $(document).on("click", ".website_sale_filter_brand_tag_show_more", function (e) {
```
Comment (GabbasovDinar) [inline website_sale_product_brand_filter/static/src/js/website_sale_product_brand_filter.js]: you need use publicWidget

**Review by GabbasovDinar — CHANGES_REQUESTED** [review]

**Review by FrancescoBallerini — APPROVED** [review]
Functional review: all good. 

Tested on 2 DB:
- PR runbot with demo data 
- personal DB with 14'000 products 

Code:

I see that we can filter by attributes as well, I am not sure if it's worth reporting, but on non-demo DB I found out by chance that [this line](https://github.com/cetmix/oca-e-commerce/blob/cfb11fe495fb1dbddfd1d0923def688276519c6a/website_sale_product_brand_filter/controllers/website_sale.py#L110) 

`"attr_valid": search_product.mapped("attribute_line_ids").mapped("value_ids").ids,`

should be monitored if you have a LOT of values mapped for one or more attributes. 

I stumbled accross it because we erroneously imported a lot of values for an attribute (i'm talking in the order of thousands) and despite this is for sure an issue by our side, I want to report that this line also slowed down the e-commerce page loading by a few seconds, which wasn't an issue in a previous version of this PR, before that line was added. 

I don't think that it will impact in a real production environment though, where you might have just a few tens or maybe hundreds of values mapped in your attributes. I can give you a feedback once we have normalized our DB attribute values, if you are interested into.

**Comment by dessanhemrayev** [comment]
> Functional review: all good.
> 
> Tested on 2 DB:
> 
> * PR runbot with demo data
> * personal DB with 14'000 products
> 
> Code:
> 
> I see that we can filter by attributes as well, I am not sure if it's worth reporting, but on non-demo DB I found out by chance that [this line](https://github.com/cetmix/oca-e-commerce/blob/cfb11fe495fb1dbddfd1d0923def688276519c6a/website_sale_product_brand_filter/controllers/website_sale.py#L110)
> 
> `"attr_valid": search_product.mapped("attribute_line_ids").mapped("value_ids").ids,`
> 
> should be monitored if you have a LOT of values mapped for one or more attributes.
> 
> I stumbled accross it because we erroneously imported a lot of values for an attribute (i'm talking in the order of thousands) and despite this is for sure an issue by our side, I want to report that this line also slowed down the e-commerce page loading by a few seconds, which wasn't an issue in a previous version of this PR, before that line was added.
> 
> I don't think that it will impact in a real production environment though, where you might have just a few tens or maybe hundreds of values mapped in your attributes. I can give you a feedback once we have normalized our DB attribute values, if you are interested into.

Thanks for the detailed comment, we will keep an eye on it).

File: website_sale_product_brand_filter/controllers/website_sale.py (hunk 3 above)
Comment (dessanhemrayev) [reply website_sale_product_brand_filter/controllers/website_sale.py]: Okay, thanks)

File: website_sale_product_brand_filter/controllers/website_sale.py (hunk 4 above)
Comment (dessanhemrayev) [reply website_sale_product_brand_filter/controllers/website_sale.py]: Okay, I'll fix it

File: website_sale_product_brand_filter/controllers/website_sale.py (hunk 5 above)
Comment (dessanhemrayev) [reply website_sale_product_brand_filter/controllers/website_sale.py]: Okay, I'll fix it

File: website_sale_product_brand_filter/controllers/website_sale.py (hunk 3 above)
Comment (dessanhemrayev) [reply website_sale_product_brand_filter/controllers/website_sale.py:77]: Okay, I'll fix it

File: website_sale_product_brand_filter/__manifest__.py (hunk 1 above)
Comment (dessanhemrayev) [reply website_sale_product_brand_filter/__manifest__.py]: Okay, I'll fix it

File: website_sale_product_brand_filter/controllers/website_sale.py (hunk 6 above)
Comment (dessanhemrayev) [reply website_sale_product_brand_filter/controllers/website_sale.py]: Thanks for the hint)

File: website_sale_product_brand_filter/controllers/website_sale.py (hunk 7 above)
Comment (dessanhemrayev) [reply website_sale_product_brand_filter/controllers/website_sale.py]: Okay, I'll fix it

File: website_sale_product_brand_filter/static/src/js/website_sale_product_brand_filter.js (hunk 9 above)
Comment (dessanhemrayev) [reply website_sale_product_brand_filter/static/src/js/website_sale_product_brand_filter.js]: Thanks for the hint)

File: website_sale_product_brand_filter/controllers/website_sale.py [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+        brands = self._remove_extra_brands(brands, search_products, attrib_values)
+        # assign values for usage in qweb
+        res.qcontext.update(
+            {
+                "brands": brands,
+                "selected_brand_ids": selected_brand_ids,
+                "attr_valid": search_products.mapped("attribute_line_ids")
+                .mapped("value_ids")
+                .ids,
```
Comment (GabbasovDinar) [inline website_sale_product_brand_filter/controllers/website_sale.py]: you can use one mapped method: e.g.

search_products.mapped("attribute_line_ids.value_ids").ids

File: website_sale_product_brand_filter/controllers/website_sale.py [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+        keep = QueryURL(
+            "/shop",
+            category=category and int(category),
+            search=search,
+            attrib=attrib_list,
+            order=post.get("order"),
+            brand=brands_list,
+            brand_ids=selected_brand_ids,
+        )
+        res.qcontext["keep"] = keep
```
Comment (GabbasovDinar) [inline website_sale_product_brand_filter/controllers/website_sale.py]:
```python
res.qcontext.update({
    "brands": brands,
    ...
    "keep": keep
})

you can use one `update` method

File: website_sale_product_brand_filter/static/src/js/website_sale_product_brand_filter.js [hunk 12]
```diff
@@ -0,0 +1,41 @@
...
+                // Hide 'Show less'
+                var $elem = $(e.currentTarget);
+                $elem.addClass("d-none");
+                // Add scroll to brands list
+                $(".website_sale_filter_brand_tags").addClass("scroll");
+                // Show 'Show more'
+                $(".website_sale_filter_brand_tag_show_more").removeClass("d-none");
+            },
+        }
+    );
```
Comment (GabbasovDinar) [inline website_sale_product_brand_filter/static/src/js/website_sale_product_brand_filter.js]: please, check my suggestion

```suggestion
    publicWidget.registry.WebsiteSaleProductBrandFilter = publicWidget.Widget.extend(
        {
            selector: "#wsale_products_attributes_collapse",
            events: {
                'click .website_sale_filter_brand_tag_show_more': '_onClickShowMore',
                'click .website_sale_filter_brand_tag_show_less': '_onClickShowLess',
            },
            _onClickShowMore: function (event) {
                event.preventDefault();
                this._switchState($(event.currentTarget), false);
                // Show 'Show less'
                this.$(".website_sale_filter_brand_tag_show_less").removeClass("d-none");
            },
             _onClickShowLess: function (event) {
                event.preventDefault();
                this._switchState($(event.currentTarget), true);
                // Show 'Show more'
                this.$(".website_sale_filter_brand_tag_show_more").removeClass("d-none");
            },
            _switchState($element, scroll) {
                $element.addClass("d-none");
                if (scroll) {
                    this.$(".website_sale_filter_brand_tags").addClass("scroll");
                } else {
                    this.$(".website_sale_filter_brand_tags").removeClass("scroll");
                }
            }
        }
    );
```

**Review by GabbasovDinar — CHANGES_REQUESTED** [review]
Thanks! Please check my some more changes

File: website_sale_product_brand_filter/controllers/website_sale.py (hunk 10 above)
Comment (dessanhemrayev) [reply website_sale_product_brand_filter/controllers/website_sale.py]: Thanks for the hint)

File: website_sale_product_brand_filter/controllers/website_sale.py (hunk 11 above)
Comment (dessanhemrayev) [reply website_sale_product_brand_filter/controllers/website_sale.py]: Thanks for the hint)

File: website_sale_product_brand_filter/static/src/js/website_sale_product_brand_filter.js (hunk 12 above)
Comment (dessanhemrayev) [reply website_sale_product_brand_filter/static/src/js/website_sale_product_brand_filter.js]: Thanks for the hint)

**Review by GabbasovDinar — APPROVED** [review]
Code LGTM

**Comment by francesco-ooops** [comment]
@dessanhemrayev squash?

**Comment by francesco-ooops** [comment]
@OCA/e-commerce-maintainers WDYT?

**Comment by francesco-ooops** [comment]
@OCA/e-commerce-maintainers good for merge?

**Comment by francesco-ooops** [comment]
@rafaelbn any opinion on this? thanks!

**Comment by francesco-ooops** [comment]
@pedrobaeza what do you say?

**Comment by pedrobaeza** [comment]
About the name, following conventions of `area` + `subarea` + function, it should `website_sale_filter_product_brand`, similar to `website_sale_filter_price`. Is it a problem for you to rename it?

**Comment by francesco-ooops** [comment]
> About the name, following conventions of `area` + `subarea` + function, it should `website_sale_filter_product_brand`, similar to `website_sale_filter_price`. Is it a problem for you to rename it?

No problem, we chose this technical name as the module depends on website_sale_product_brand, so user can easily find both

website_sale_product_brand
website_sale_product_brand_filter

but if you confirm it should be renamed we will proceed :)

**Comment by pedrobaeza** [comment]
Well, it's not critical, and yes, there's no absolute truth here. It's OK then to keep it if there's such dependency, but I don't quite understand why you need to depend on `website_sale_product_brand`, as they seems different and barely they can share code.

**Comment by dessanhemrayev** [comment]
> Well, it's not critical, and yes, there's no absolute truth here. It's OK then to keep it if there's such dependency, but I don't quite understand why you need to depend on `website_sale_product_brand`, as they seems different and barely they can share code.

Okay, we will fix it)

File: website_sale_product_brand_filter/readme/README.rst
Comment (pedrobaeza) [inline website_sale_product_brand_filter/readme/README.rst:1]: Remove these empty files

**Review by pedrobaeza — APPROVED** [review]
Thanks for the changes 

/ocabot merge nobump

**Pull request merged**